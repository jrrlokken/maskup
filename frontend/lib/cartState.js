import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

const CartStateProvider = ({ children }) => {
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = () => {
    setCartOpen(!cartOpen);
  }
  const closeCart = () => {
    setCartOpen(false);
  }
  const openCart = () => {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{
        cartOpen,
        setCartOpen,
        toggleCart,
        closeCart,
        openCart
      }}
    >
      {children}
    </LocalStateProvider>
  );
}

const useCart = () => {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };