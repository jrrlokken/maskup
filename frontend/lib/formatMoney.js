export default function formatMoney(amount) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  };
  const formatter = Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}