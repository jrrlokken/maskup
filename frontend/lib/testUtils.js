import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const type = (wrapper, name, value) => {
  wrapper.find(`input[name='${name}']`).simulate('change', {
    target: { name, value }
  });
};

const fakeProduct = ({
  // __typename: 'Item',
  id: 'abc123',
  name: 'Test Product',
  description: 'This is a test product',
  price: 1499,
  photo: {
    altText: 'dog',
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
  },
  user: null,
});

const fakeUser = overrides => ({
  // __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  // __typename: 'OrderItem',
  id: casual.uuid,
  photo: {
    image: {
      publicUrlTransformed: `${casual.word}.jpg`
    }
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  // __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = overrides => ({
  // __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeProduct,
  user: fakeUser(),
  ...overrides,
});


// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  type,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  LocalStorageMock,
};