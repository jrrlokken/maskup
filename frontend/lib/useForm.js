import { useState, useEffect } from 'react';

const useForm = (initial = {}) => {
  const [inputs, setInputs] = useState(initial);
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    setInputs(initial);
  }, [initialValues]);

  const handleChange = event => {
    let { value, name, type } = event.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = event.target.files;
    }
    setInputs({
      ...inputs,
      [name]: value,
    });
  }

  const resetForm = () => {
    setInputs(initial);
  }

  const clearForm = () => {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm
  };
}

export default useForm;