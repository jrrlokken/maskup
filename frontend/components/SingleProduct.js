import { useQuery } from '@apollo/client';
import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import styled from 'styled-components';

import DisplayError from './ErrorMessage';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import DeleteProduct from './DeleteProduct';
import AddToCart from './AddToCart';
import formatMoney from '../lib/formatMoney';

const ProductStyles = styled.div`
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  max-width: var(--maxWidth);
  justify-content: center;
  align-items: top;
  gap: 2rem;
  img {
    width: 100%;
    object-fit: contain;
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    Product(where: { id: $id }) {
      id
      name
      description
      price
      photo {
        altText
        image {
          publicUrlTransformed
        }
      }
    }
  }
`;

const SingleProduct = ({ id }) => {
  const { data, error, loading } = useQuery(SINGLE_ITEM_QUERY, {
    variables: {
      id
    }
  });
  if (loading) return <p>Loading...</p>;
  if (error) return <DisplayError error={error} />;
  const { Product } = data;

  return (
    <ItemStyles>
      <img
        src={Product.photo.image.publicUrlTransformed}
        alt={Product.name}
      />
      <Title>
        <Link href={`/product/${Product.id}`}>
          {Product.name}
        </Link>
      </Title>
      <PriceTag>{formatMoney(Product.price)}</PriceTag>
      <p>{Product.description}</p>
      <div className="buttonList">
        <Link
          href={{
            pathname: 'update',
            query: { id: Product.id },
          }}
        >
          Edit ✏️
        </Link>
        <AddToCart id={Product.id} />
        <DeleteProduct id={Product.id}>Delete</DeleteProduct>
      </div>
    </ItemStyles>
  );
}

export default SingleProduct;
export { SINGLE_ITEM_QUERY };