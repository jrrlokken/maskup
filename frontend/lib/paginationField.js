import { PAGINATION_QUERY } from '../components/Pagination';

const paginationField = () => {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing
        .slice(skip, skip + first)
        .filter((x) => x);

      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return;
      }
      if (items.length) {
        return items;
      }
      return false;
    },

    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}

export default paginationField;