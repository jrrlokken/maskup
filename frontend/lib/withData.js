import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/react-ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';

import { endpoint, productionEndpoint } from '../config';
import paginationField from './paginationField';

const createClient = ({ headers, initialState }) => {
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),

      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : productionEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allProducts: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
