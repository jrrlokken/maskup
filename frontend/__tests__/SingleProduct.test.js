import { configure, mount } from 'enzyme';
import Adapter from '@hteker/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from '@apollo/client/testing';
import { act } from 'react-dom/test-utils';
import wait from 'waait';

import SingleProduct, { SINGLE_ITEM_QUERY } from '../components/SingleProduct';
import AddToCart from '../components/AddToCart';
import { fakeProduct } from '../lib/testUtils';

configure({ adapter: new Adapter() });

const mocks = [
  {
    request: { 
      query: SINGLE_ITEM_QUERY,
      variables: { id: fakeProduct.id }
    },
    result: {
      data: { Product: fakeProduct }
    }
  }
];

describe('<SingleProduct/>', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SingleProduct id={fakeProduct.id} />
      </MockedProvider>
    );
    expect(wrapper.find('p').text()).toBe('Loading...');
    await act(async () => {
      await wait();
      wrapper.update();
      console.log(wrapper.debug());
      expect(wrapper.find('Title').text()).toBe('Test Product');
      expect(wrapper.find('PriceTag').text()).toBe('$14.99');
      expect(wrapper.containsMatchingElement(<AddToCart />)).toBe(true);
      const component = wrapper.find('SingleProduct');
      expect(toJSON(component)).toMatchSnapshot();
    });
  });
})

