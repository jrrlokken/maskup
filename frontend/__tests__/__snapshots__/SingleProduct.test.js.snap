// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct/> renders and matches snapshot 1`] = `
<SingleProduct
  id="abc123"
>
  <ItemStyles__Item>
    <div
      className="ItemStyles__Item-sc-1cqsy29-0 bnaCVl"
    >
      <img
        alt="Test Product"
        src="dog.jpg"
      />
      <Title>
        <h3
          className="Title-sc-61hg5m-0 kkIpao"
        >
          <Link
            href="/product/abc123"
          >
            <a
              href="/product/abc123"
              onClick={[Function]}
              onMouseEnter={[Function]}
            >
              Test Product
            </a>
          </Link>
        </h3>
      </Title>
      <PriceTag>
        <span
          className="PriceTag-sc-17vnhm5-0 fLBUwy"
        >
          $14.99
        </span>
      </PriceTag>
      <p>
        This is a test product
      </p>
      <div
        className="buttonList"
      >
        <Link
          href={
            Object {
              "pathname": "update",
              "query": Object {
                "id": "abc123",
              },
            }
          }
        >
          <a
            href="/update?id=abc123"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            Edit ‚úèÔ∏è
          </a>
        </Link>
        <AddToCart
          id="abc123"
        >
          <button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Add
             To Cart üõí
          </button>
        </AddToCart>
        <DeleteProduct
          id="abc123"
        >
          <button
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </DeleteProduct>
      </div>
    </div>
  </ItemStyles__Item>
</SingleProduct>
`;
